runtime/sema/check_casting_expression.go

```diff
@@ -412,3 +421,204 @@ func FailableCastCanSucceed(subType, superType Type) bool {
 
 	return true
 }
+
+// IsCastRedundant checks whether a simple cast is redundant.
+// Checks for two cases:
+//    - Case I: Contextually expected type is same as casted type (target type).
+//    - Case II: Expression is self typed, and is same as the casted type (target type).
+func IsCastRedundant(expr ast.Expression, exprInferredType, targetType, expectedType Type) bool {
+	if expectedType != nil && expectedType.Equal(targetType) {
+		return true
+	}
+
+	checkCastVisitor := &CheckCastVisitor{}
+
+	return checkCastVisitor.isCastRedundant(expr, exprInferredType, targetType)
+}
+
+type CheckCastVisitor struct {
+	exprInferredType Type
+	targetType       Type
+}
+
+var _ ast.ExpressionVisitor = &CheckCastVisitor{}
```

Comment:
Not quite sure, but could a simpler solution like be sufficient here: Would it be enough to just infer the type of the expression without the expected type, and compare it to the casted type (target type)?

Basically, in `IsCastRedundant`:

```go
if expectedType == nil {
	expectedType = // ... infer type for casted expression
} 
return expectedType.Equal(targetType)
```

Reply:
The problem with that is, sometimes the casted/target type is used as the annotation to infer the type for the expression. e.g: `[1, 3] as [UInt8]` Then it would en

So we could have to run one pass with the expected type, and another pass without the expected type and compare the results of the two. But two passes would be even more expensive (`CheckCastVIsitor` doesn't recursively visit all the nested expressions, so it's fairly less expensive), also we would need a way to disable the error reporting in the second pass.

Reply:
Makes sense 👍